### PR TITLE
WT-12508 Add code complexity and per function info to code change report

### PR DESCRIPTION
**Pull request by jeremythorp**
Added code complexity and per function info to the code change report, including the delta to the previous version (for post-merge tests) and the branch point from develop (for patch builds), along with uncovered complexity.
Added an overall summary for the changed lines.

### CODE REVIEW

File: test/evergreen/code_change_report/code_change_info.py
Comment (ajmorton) [inline test/evergreen/code_change_report/code_change_info.py:348]: Do these arguments need to be optional? It'd be simpler to require them and not need to check if they're available

File: test/evergreen/code_change_report/code_change_report.py
Comment (ajmorton) [inline test/evergreen/code_change_report/code_change_report.py:307]:
```suggestion
# coverage_string converts a coverage value and a total into a colour-coded html string
```

Comment (ajmorton) [reply test/evergreen/code_change_report/code_change_report.py:307]: (Technically `colour` should also become `color`, but we're not running s_all on these scripts so I'm happy to leave as is)

**Review by ajmorton — COMMENTED** [review]
A couple of comments and questions to kick things off, but I don't see any major changes required 👍

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:247]: Well spotted, at least the typo was consistent :-).

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:306]: This is an artefact of format of the diff information. If a line is simply moved (eg because something is inserted or deleted before it) then it will have an 'old' line number. On the other hand, if a line is changed, then you see it as a line being deleted, and a line being added. The new version of the line will not have an 'old' line number. You can see this happening in this Code Change Report at line 850 in conn_log.c: https://build_external.s3.amazonaws.com/wiredtiger/code-statistics/caa512a029b5486c94fc6eb8a0070c66d1948558/code-change-report_wiredtiger_code_statistics_caa512a029b5486c94fc6eb8a0070c66d1948558_24_02_28_22_21_56-0/code_change_report.html
I'll add more comments to make that clearer.

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:348]: That makes sense. During the development, I didn't always have those files available, and so I made them optional. But now the work is (nearly) complete, they should always be there.

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:313]: Well spotted on the typo!

Rather than removing the get_num_branches_covered function, I've replaced it's implementation with your (simpler) expression.

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:306]: Done.

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py]: Now fixed.

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py]: I've made those options required.

File: test/evergreen.yml
Comment (jeremythorp) [reply test/evergreen.yml:3061]: Thanks for the suggestion - I've made that change.

File: test/evergreen.yml
Comment (jeremythorp) [reply test/evergreen.yml:3046]: I absolutely agree we should move as much out of evergreen.yml as possible. That's part of the reason that the majority of the code change report work is performed by the Python scripts. I'd like to move the shell script out down the track, but it's made a bit complicated by the reliance on a number of Evergreen variables (eg is_patch and  github_commit), so the script would need some re-writing.

Comment (jeremythorp) [reply test/evergreen.yml:3046]: In fact, it's worth considering converting the shell script into a Python script.

File: test/evergreen/download_metrixpp.sh
Comment (raviprakashgiri29) [inline test/evergreen/download_metrixpp.sh:8]: Any reason to checkout a particular commit ? I believe earlier we were just checking out the latest metrixplusplus.

File: test/evergreen/code_change_report/code_change_info.py
Comment (ajmorton) [inline test/evergreen/code_change_report/code_change_info.py:304]:
```suggestion
                    # Changed lines of code appear as two entries: (1) a deleted line and (2) an added line.
```

File: test/evergreen/code_change_report/code_change_info.py
Comment (ajmorton) [inline test/evergreen/code_change_report/code_change_info.py:257]: This is a typo from my original suggestion. The `> 0` needs to be outside the `len` call:
```suggestion
    return len([x for x in branch_coverage_info if x['count']])
```
```suggestion
    return len([x for x in branch_coverage_info if x['count'] > 0]) > 0
```

**Review by ajmorton — APPROVED** [review]
LGTM with one code change. No need for re-review 🎉
Since this is a large change I'd prefer if we could get a second LGTM before merging.

File: test/evergreen/download_metrixpp.sh
Comment (jeremythorp) [reply test/evergreen/download_metrixpp.sh:8]: Hi Ravi, I'm keen that we switch to a specific version of Metric++ so that we can control upgrades (including if they have any issues). Metrix++ doesn't have version numbers or releases, so a git hash is the next best thing. 
I'll expand the comment to make this clearer for future readers.

File: test/evergreen.yml
Comment (jeremythorp) [reply test/evergreen.yml:3046]: I'll create a seperate ticket to refactor the shell script into a Python script.

File: test/evergreen/code_change_report/code_change_info.py
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py]: Hi @ajmorton , I believe the original code is correct (ie with the `>0` inside the `len` call) because we want to count the number of branches that have an execution count greater than 0.
I've added an extra comment and split the code up to (hopefully) make it easier to understand.

**Review by raviprakashgiri29 — APPROVED** [review]
LGTM

File: test/evergreen/code_change_report/code_change_info.py
Comment (ajmorton) [reply test/evergreen/code_change_report/code_change_info.py]: Ah yep my bad. I had it in my head we were returning a bool not an int.
Resolving

**Review by ajmorton — APPROVED** [review]

**Pull request merged**
